File: app/src/main/java/life/plank/juna/zone/view/fragment/league/LeagueInfoFragment.java
package life.plank.juna.zone.view.fragment.league;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.google.gson.Gson;
import com.squareup.picasso.Picasso;

import java.net.HttpURLConnection;
import java.util.List;
import java.util.Objects;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import life.plank.juna.zone.R;
import life.plank.juna.zone.ZoneApplication;
import life.plank.juna.zone.data.network.interfaces.RestApi;
import life.plank.juna.zone.data.network.model.PlayerStatsModel;
import life.plank.juna.zone.data.network.model.StandingModel;
import life.plank.juna.zone.data.network.model.TeamStatsModel;
import life.plank.juna.zone.view.activity.FixtureActivity;
import life.plank.juna.zone.view.activity.MatchResultActivity;
import life.plank.juna.zone.view.activity.MatchResultDetailActivity;
import life.plank.juna.zone.view.adapter.PlayerStatsAdapter;
import life.plank.juna.zone.view.adapter.StandingTableAdapter;
import life.plank.juna.zone.view.adapter.TeamStatsAdapter;
import retrofit2.Response;
import rx.Observer;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static life.plank.juna.zone.util.AppConstants.PLAYER_STATS;
import static life.plank.juna.zone.util.AppConstants.STANDINGS;
import static life.plank.juna.zone.util.AppConstants.TEAM_STATS;

public class LeagueInfoFragment extends Fragment {

    private static final String TAG = LeagueInfoFragment.class.getSimpleName();

    @BindView(R.id.match_fixture_result)
    CardView matchFixtureResultLayout;
    @BindView(R.id.standing_recycler_view)
    RecyclerView standingRecyclerView;
    @BindView(R.id.team_stats_recycler_view)
    RecyclerView teamStatsRecyclerView;
    @BindView(R.id.player_stats_recycler_view)
    RecyclerView playerStatsRecyclerView;
    @BindView(R.id.progress_bar)
    ProgressBar progressBar;
    @BindView(R.id.no_standings)
    TextView noStandingsTextView;
    @BindView(R.id.no_team_stats)
    TextView noTeamStatsTextView;
    @BindView(R.id.no_player_stats)
    TextView noPlayerStatsTextView;
    @BindView(R.id.see_all_fixtures)
    TextView seeAllFixtures;
    @BindView(R.id.see_all_standings)
    TextView seeAllStandings;
    @BindView(R.id.see_more_team_stats)
    TextView seeMoreTeamStats;
    @BindView(R.id.see_more_player_stats)
    TextView seeMorePlayerStats;
    @BindView(R.id.standings_layout)
    CardView standingsLayout;
    @BindView(R.id.team_stats_layout)
    CardView teamStatsLayout;
    @BindView(R.id.player_stats_layout)
    CardView playerStatsLayout;

    @Inject
    @Named("footballData")
    RestApi restApi;
    @Inject
    Picasso picasso;
    @Inject
    Gson gson;

    private StandingTableAdapter standingTableAdapter;
    private PlayerStatsAdapter playerStatsAdapter;
    private TeamStatsAdapter teamStatsAdapter;
    private String seasonName;
    private String leagueName;
    private String countryName;

    public LeagueInfoFragment() {
    }

    public static LeagueInfoFragment newInstance(Context context, String seasonName, String leagueName, String countryName) {
        LeagueInfoFragment fragment = new LeagueInfoFragment();
        Bundle args = new Bundle();
        args.putString(context.getString(R.string.season_name), seasonName);
        args.putString(context.getString(R.string.league_name), leagueName);
        args.putString(context.getString(R.string.country_name), countryName);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args != null) {
            seasonName = args.getString(getString(R.string.season_name));
            leagueName = args.getString(getString(R.string.league_name));
            countryName = args.getString(getString(R.string.country_name));
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_league_info, container, false);
        ButterKnife.bind(this, rootView);

        ZoneApplication.getApplication().getUiComponent().inject(this);
        prepareRecyclerViews();
        getStandings();
        getPlayerStats();
        getTeamStats();

        return rootView;
    }

    private void prepareRecyclerViews() {
        standingTableAdapter = new StandingTableAdapter(picasso);
        standingRecyclerView.setAdapter(standingTableAdapter);

        playerStatsAdapter = new PlayerStatsAdapter();
        playerStatsRecyclerView.setAdapter(playerStatsAdapter);

        teamStatsAdapter = new TeamStatsAdapter(picasso);
        teamStatsRecyclerView.setAdapter(teamStatsAdapter);
    }

    public void getStandings() {
        progressBar.setVisibility(View.VISIBLE);
        restApi.getStandings(leagueName, seasonName, countryName)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Response<List<StandingModel>>>() {
                    @Override
                    public void onCompleted() {
                        Log.i(TAG, "onCompleted: getStandings()");
                    }

                    @Override
                    public void onError(Throwable e) {
                        updateUI(false, standingRecyclerView, seeAllStandings, noStandingsTextView);
                        Log.e(TAG, "onError: " + e);
                    }

                    @Override
                    public void onNext(Response<List<StandingModel>> response) {
                        switch (response.code()) {
                            case HttpURLConnection.HTTP_OK:
                                updateUI(true, standingRecyclerView, seeAllStandings, noStandingsTextView);
                                standingTableAdapter.update(response.body());
                                break;
                            case HttpURLConnection.HTTP_NOT_FOUND:
                                updateUI(false, standingRecyclerView, seeAllStandings, noStandingsTextView);
                            default:
                                updateUI(false, standingRecyclerView, seeAllStandings, noStandingsTextView);
                                break;
                        }
                    }
                });
    }

    public void getPlayerStats() {
        restApi.getPlayerStats(leagueName, seasonName, countryName)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Response<List<PlayerStatsModel>>>() {
                    @Override
                    public void onCompleted() {
                        Log.i(TAG, "onCompleted: getPlayerStats()");
                    }

                    @Override
                    public void onError(Throwable e) {
                        updateUI(false, playerStatsRecyclerView, seeMorePlayerStats, noPlayerStatsTextView);
                        Log.e(TAG, " Error" + e);
                    }

                    @Override
                    public void onNext(Response<List<PlayerStatsModel>> response) {
                        switch (response.code()) {
                            case HttpURLConnection.HTTP_OK:
                                updateUI(true, playerStatsRecyclerView, seeMorePlayerStats, noPlayerStatsTextView);
                                playerStatsAdapter.update(response.body());
                                break;
                            case HttpURLConnection.HTTP_NOT_FOUND:
                                updateUI(false, playerStatsRecyclerView, seeMorePlayerStats, noPlayerStatsTextView);
                                break;
                            default:
                                updateUI(false, playerStatsRecyclerView, seeMorePlayerStats, noPlayerStatsTextView);
                                break;
                        }
                    }
                });
    }

    public void getTeamStats() {
        progressBar.setVisibility(View.VISIBLE);
        restApi.getTeamStats(leagueName, seasonName, countryName)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Response<List<TeamStatsModel>>>() {
                    @Override
                    public void onCompleted() {
                        Log.i(TAG, "onCompleted: getTeamStats()");
                    }

                    @Override
                    public void onError(Throwable e) {
                        updateUI(false, teamStatsRecyclerView, seeMoreTeamStats, noTeamStatsTextView);
                        Log.e(TAG, " Error: " + e);
                    }

                    @Override
                    public void onNext(Response<List<TeamStatsModel>> response) {
                        switch (response.code()) {
                            case HttpURLConnection.HTTP_OK:
                                updateUI(true, teamStatsRecyclerView, seeMoreTeamStats, noTeamStatsTextView);
                                teamStatsAdapter.update(response.body());
                                break;
                            case HttpURLConnection.HTTP_NOT_FOUND:
                                updateUI(false, teamStatsRecyclerView, seeMoreTeamStats, noTeamStatsTextView);
                                break;
                            default:
                                updateUI(false, teamStatsRecyclerView, seeMoreTeamStats, noTeamStatsTextView);
                                break;
                        }
                    }
                });
    }

    private void updateUI(boolean available, RecyclerView recyclerView, TextView seeMoreView, TextView noDataView) {
        progressBar.setVisibility(View.INVISIBLE);
        recyclerView.setVisibility(available ? View.VISIBLE : View.INVISIBLE);
        seeMoreView.setVisibility(available ? View.VISIBLE : View.GONE);
        noDataView.setVisibility(available ? View.GONE : View.VISIBLE);
    }

    @OnClick({R.id.see_all_fixtures, R.id.see_all_standings, R.id.see_more_team_stats, R.id.see_more_player_stats})
    public void onItemClick(View view) {
        ((MatchResultActivity) Objects.requireNonNull(getActivity())).updateBackgroundBitmap();
        switch (view.getId()) {
            case R.id.see_all_fixtures:
                FixtureActivity.launch(
                        getActivity(),
                        seasonName,
                        leagueName,
                        countryName,
                        matchFixtureResultLayout
                );
                break;
            case R.id.see_all_standings:
                MatchResultDetailActivity.launch(getActivity(), STANDINGS, seasonName, leagueName, countryName,
                        gson.toJson(standingTableAdapter.getStandings()), standingsLayout);
                break;
            case R.id.see_more_team_stats:
                MatchResultDetailActivity.launch(getActivity(), TEAM_STATS, seasonName, leagueName, countryName,
                        gson.toJson(teamStatsAdapter.getTeamStats()), teamStatsLayout);
                break;
            case R.id.see_more_player_stats:
                MatchResultDetailActivity.launch(getActivity(), PLAYER_STATS, seasonName, leagueName, countryName,
                        gson.toJson(playerStatsAdapter.getPlayerStats()), playerStatsLayout);
                break;
        }
    }

    @Override
    public void onDestroy() {
        standingTableAdapter = null;
        teamStatsAdapter = null;
        playerStatsAdapter = null;
        super.onDestroy();
    }
}

File: app/src/main/java/life/plank/juna/zone/view/fragment/LeagueInfoFragment.kt
package life.plank.juna.zone.view.fragment


import android.arch.lifecycle.Observer
import android.os.AsyncTask
import android.os.Bundle
import android.os.Parcelable
import android.support.v7.widget.CardView
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.bumptech.glide.Glide
import com.google.gson.Gson
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.fragment_league_info.*
import kotlinx.android.synthetic.main.item_standings.*
import kotlinx.android.synthetic.main.layout_league_info.*
import kotlinx.android.synthetic.main.league_toolbar.*
import life.plank.juna.zone.R
import life.plank.juna.zone.ZoneApplication
import life.plank.juna.zone.data.local.model.LeagueInfo
import life.plank.juna.zone.data.model.*
import life.plank.juna.zone.data.network.interfaces.RestApi
import life.plank.juna.zone.util.AppConstants.*
import life.plank.juna.zone.util.DataUtil.isNullOrEmpty
import life.plank.juna.zone.util.DateUtil.getDateDiffFromToday
import life.plank.juna.zone.util.hideAndShowBoomMenu
import life.plank.juna.zone.util.setupBoomMenu
import life.plank.juna.zone.view.activity.FixtureActivity
import life.plank.juna.zone.view.activity.LeagueInfoActivity
import life.plank.juna.zone.view.activity.LeagueInfoActivity.fixtureByMatchDayList
import life.plank.juna.zone.view.activity.LeagueInfoDetailActivity
import life.plank.juna.zone.view.adapter.FixtureAdapter
import life.plank.juna.zone.view.adapter.PlayerStatsAdapter
import life.plank.juna.zone.view.adapter.StandingTableAdapter
import life.plank.juna.zone.view.adapter.TeamStatsAdapter
import life.plank.juna.zone.view.fragment.base.BaseLeagueFragment
import org.jetbrains.anko.sdk27.coroutines.onClick
import java.lang.ref.WeakReference
import java.util.*
import javax.inject.Inject
import javax.inject.Named

class LeagueInfoFragment : BaseLeagueFragment() {

    lateinit var fixtureByMatchDayList: MutableList<FixtureByMatchDay>

    @Inject
    lateinit var picasso: Picasso
    @Inject
    lateinit var gson: Gson
    @field: [Inject Named("footballData")]
    lateinit var restApi: RestApi
    private var fixtureAdapter: FixtureAdapter? = null
    private var standingTableAdapter: StandingTableAdapter? = null
    private var playerStatsAdapter: PlayerStatsAdapter? = null
    private var teamStatsAdapter: TeamStatsAdapter? = null
    private var isDataLocal: Boolean = false
    private lateinit var league: League

    companion object {
        private var TAG = LeagueInfoActivity::class.java.simpleName
        fun newInstance(league: League) = LeagueInfoFragment().apply { arguments = Bundle().apply { putParcelable(getString(R.string.intent_league), league) } }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ZoneApplication.getApplication().uiComponent.inject(this)
        arguments?.apply { league = getParcelable(getString(R.string.intent_league))!! }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
            inflater.inflate(R.layout.fragment_league_info, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupBoomMenu(BoomMenuPage.BOOM_MENU_SETTINGS_AND_HOME, activity!!, null, arc_menu, null)

        prepareRecyclerViews()
        getLeagueInfoFromRoomDb()

        title.text = league.name
        logo.setImageResource(league.leagueLogo)
        root_layout.setBackgroundColor(resources.getColor(league.dominantColor!!, null))
        hideAndShowBoomMenu(nestedScrollView, arc_menu)

        setOnClickListeners()
    }

    private fun setOnClickListeners() {
        see_all_fixtures.onClick {
            if (!isNullOrEmpty<FixtureByMatchDay>(fixtureByMatchDayList)) {
                FixtureActivity.launch(activity, league)
            }
        }
        see_all_standings.onClick {
            LeagueInfoDetailActivity.launch(activity, STANDINGS, standingTableAdapter!!.standings as ArrayList<out Parcelable>, standings_layout)
        }
        see_more_team_stats.onClick {
            LeagueInfoDetailActivity.launch(activity, TEAM_STATS, teamStatsAdapter!!.teamStats as ArrayList<out Parcelable>, team_stats_layout)
        }
        see_more_player_stats.onClick {
            LeagueInfoDetailActivity.launch(activity, PLAYER_STATS, playerStatsAdapter!!.playerStats as ArrayList<out Parcelable>, player_stats_layout)
        }
    }

    override fun getRootFadedCardLayout(): ViewGroup = faded_card_layout

    override fun getRootCard(): CardView = root_card

    override fun getDragHandle(): View = drag_area

    override fun getGlide() = Glide.with(this)

    override fun getTheGson() = gson

    override fun getTheLeague() = league

    private fun prepareRecyclerViews() {
//        fixtureAdapter = FixtureAdapter(null, activity)
        fixtures_section_list.adapter = fixtureAdapter

        standingTableAdapter = StandingTableAdapter(picasso)
        standing_recycler_view.adapter = standingTableAdapter

        playerStatsAdapter = PlayerStatsAdapter()
        player_stats_recycler_view.adapter = playerStatsAdapter

        teamStatsAdapter = TeamStatsAdapter(picasso)
        team_stats_recycler_view.adapter = teamStatsAdapter
    }

    private fun getLeagueInfoFromRoomDb() {
        isDataLocal = true
        leagueViewModel.leagueInfoLiveData.observe(this, Observer<LeagueInfo> { handleLeagueInfoData(it) })
        leagueViewModel.getLeagueInfoFromDb(league.id)
    }

    private fun getLeagueInfoFromRestApi() {
        if (isDataLocal) {
            isDataLocal = false
            leagueViewModel.getLeagueInfoFromRestApi(league, restApi)
        }
    }

    private fun handleLeagueInfoData(leagueInfo: LeagueInfo?) {
        if (leagueInfo != null) {
//            Update new data in DB
            if (!isDataLocal) {
                leagueViewModel.leagueRepository.insertLeagueInfo(leagueInfo)
            }

            fixture_progress_bar.visibility = View.GONE
            if (leagueInfo.fixtureByMatchDayList == emptyList<Any>() || isNullOrEmpty<FixtureByMatchDay>(leagueInfo.fixtureByMatchDayList)) {
                updateUI(false, fixtures_section_list, see_all_fixtures, fixture_no_data)
            } else {
                fixtureByMatchDayList = leagueInfo.fixtureByMatchDayList.toMutableList()
                UpdateFixtureAdapterTask.parse(this)
                updateUI(true, fixtures_section_list, see_all_fixtures, fixture_no_data)
            }

            standings_progress_bar.visibility = View.GONE
            if (leagueInfo.standingsList == emptyList<Any>() || isNullOrEmpty<Standings>(leagueInfo.standingsList)) {
                updateUI(false, standing_recycler_view, see_all_standings, no_standings)
            } else {
                updateUI(true, standing_recycler_view, see_all_standings, no_standings)
                standingTableAdapter!!.update(leagueInfo.standingsList)
            }

            team_stats_progress_bar.visibility = View.GONE
            if (leagueInfo.teamStatsList == emptyList<Any>() || isNullOrEmpty<TeamStats>(leagueInfo.teamStatsList)) {
                updateUI(false, team_stats_recycler_view, see_more_team_stats, no_team_stats)
            } else {
                updateUI(true, team_stats_recycler_view, see_more_team_stats, no_team_stats)
                teamStatsAdapter!!.update(leagueInfo.teamStatsList)
            }

            player_stats_progress_bar.visibility = View.GONE
            if (leagueInfo.playerStatsList == emptyList<Any>() || isNullOrEmpty<PlayerStats>(leagueInfo.playerStatsList)) {
                updateUI(false, player_stats_recycler_view, see_more_player_stats, no_player_stats)
            } else {
                updateUI(true, player_stats_recycler_view, see_more_player_stats, no_player_stats)
                playerStatsAdapter!!.update(leagueInfo.playerStatsList)
            }
            getLeagueInfoFromRestApi()
        } else {
            getLeagueInfoFromRestApi()
        }
    }

    private fun updateUI(available: Boolean, recyclerView: RecyclerView, seeMoreView: TextView, noDataView: TextView) {
        recyclerView.visibility = if (available) View.VISIBLE else View.INVISIBLE
        seeMoreView.visibility = if (available) View.VISIBLE else View.GONE
        noDataView.visibility = if (available) View.GONE else View.VISIBLE
    }

    override fun onDestroy() {
        fixtureAdapter = null
        standingTableAdapter = null
        teamStatsAdapter = null
        playerStatsAdapter = null
        if (!isNullOrEmpty<FixtureByMatchDay>(fixtureByMatchDayList)) {
            fixtureByMatchDayList.clear()
        }
        super.onDestroy()
    }

    private class UpdateFixtureAdapterTask private constructor(leagueInfoFragment: LeagueInfoFragment) : AsyncTask<Void, Void, List<MatchFixture>>() {

        private val ref: WeakReference<LeagueInfoFragment> = WeakReference(leagueInfoFragment)
        private var recyclerViewScrollIndex = 0

        companion object {
            internal fun parse(leagueInfoFragment: LeagueInfoFragment) {
                UpdateFixtureAdapterTask(leagueInfoFragment).execute()
            }
        }

        override fun onPreExecute() {
            ref.get()?.run {
                fixture_progress_bar.visibility = View.VISIBLE
                see_all_fixtures.isEnabled = false
                see_all_fixtures.isClickable = false
            }
        }

        override fun doInBackground(vararg voids: Void): List<MatchFixture>? {
            var isPastMatches = true
            if (!isNullOrEmpty<FixtureByMatchDay>(fixtureByMatchDayList)) {
                for (matchDay in fixtureByMatchDayList!!) {
                    try {
                        if (matchDay.daySection == PAST_MATCHES) {
                            isPastMatches = true
                            recyclerViewScrollIndex = fixtureByMatchDayList!!.indexOf(matchDay)
                        } else if (matchDay.daySection == TODAY_MATCHES) {
                            isPastMatches = false
                            recyclerViewScrollIndex = fixtureByMatchDayList!!.indexOf(matchDay)
                        }
                    } catch (e: Exception) {
                        Log.e("FixtureAdapterTask", "doInBackground: recyclerViewScrollIndex ", e)
                    }

                }
                val matchFixtures = ArrayList<MatchFixture>()
                getMatchesToShow(matchFixtures, isPastMatches)
                return if (matchFixtures.size >= 4) matchFixtures.subList(0, 4) else matchFixtures
            }
            return null
        }

        private fun getMatchesToShow(matchFixtures: MutableList<MatchFixture>, isPastMatches: Boolean) {
            val fixtureByDateList = fixtureByMatchDayList!![recyclerViewScrollIndex].fixtureByDateList
            for (fixtureByDate in fixtureByDateList) {
                for (matchFixture in fixtureByDate.fixtures) {
                    try {
                        if (isPastMatches && getDateDiffFromToday(matchFixture.matchStartTime) <= 0) {
                            matchFixtures.add(matchFixture)
                        } else if (getDateDiffFromToday(matchFixture.matchStartTime) <= 1) {
                            matchFixtures.add(matchFixture)
                        }
                    } catch (e: Exception) {
                        Log.e("FixtureAdapterTask", "doInBackground: getDateDiffFromToday() ", e)
                    }

                }
            }
            if (isPastMatches) {
                matchFixtures.reverse()
            }
        }

        override fun onPostExecute(matchFixtures: List<MatchFixture>) {
            ref.get()?.run {
                if (!isNullOrEmpty<MatchFixture>(matchFixtures)) {
                    if (fixtureAdapter != null) {
                        fixtureAdapter!!.update(matchFixtures)
                    }
                    fixtures_section_list.scrollToPosition(recyclerViewScrollIndex)
                    see_all_fixtures.isEnabled = true
                    see_all_fixtures.isClickable = true
                }
                fixture_progress_bar.visibility = View.GONE
            }
            if (ref.get() != null) {
            }
        }
    }
}
